app/demo-geant-integration/RunAction.cc

```diff
@@ -69,14 +69,17 @@ void RunAction::BeginOfRunAction(const G4Run* run)
 void RunAction::EndOfRunAction(const G4Run*)
 {
     CELER_LOG_LOCAL(status) << "Finalizing Celeritas";
+    celeritas::ExceptionConverter call_g4exception{"celer0005"};
 
-    // Deallocate Celeritas state data (optional)
-    transport_.reset();
+    // Deallocate Celeritas state data (ensures that objects are deleted on the
+    // thread in which they're created, necessary by some geant4
+    // thread-local allocators)
+    CELER_TRY_ELSE(transport_->Finalize(), call_g4exception);
 
     // Clear shared data and write if master thread (when running without MT)
     if (G4Threading::IsMasterThread())
     {
-        params_->Finalize();
+        CELER_TRY_ELSE(params_->Finalize(), call_g4exception);
```

Comment:
Line 77 only finalizes (clears) the local state; and 82 is necessary to reset the shared data if MT is disabled. (I don't *think* anything should break if we wait for the params to destruct on their own, but I think it's safer to do in the end-of-run action.)

Reply:
In Geant4 build, there are sequential **only** mode and sequential mode (build with MT, but equivalent to one worker thread).  I was confused as there is  a separate `MasterRunAction` (which I thought should be called for the sequential).

Reply:
The main executable creates the run manager with:
```c++
    run_manager.reset(G4RunManagerFactory::CreateRunManager(
        num_threads > 1 ? G4RunManagerType::MT : G4RunManagerType::Serial));
```
when using `G4RunManagerType::Serial`, `G4VUserActionInitialization::BuildForMaster` never gets called, so the `MasterRunAction` isn't created.

If I'm understanding correctly, you're saying if we only used `G4RunManagerType::MT` then we wouldn't have to include the `if (G4Threading::IsMasterThread())` here, because this `RunAction` is only for the worker threads. But since we do support the `Serial` run manager type, I think we do have to have it (because "worker" and "master" are the same thread in that case).

Reply:
I kind of catch that.  So, is the current approach work for both `G4RunManagerType::Serial `and `G4RunManagerType::SerialOnly` or `G4RunManagerType::MT `and `G4RunManagerType::MTOnly` cases?

Reply:
As far as I know, `SerialOnly` is the Geant4 build with GEANT4_BUILD_MULTITHREADED=OFF (i.e., native sequential) and `Serial` is the build with GEANT4_BUILD_MULTITHREADED=ON but with one thread (so, there is still a master thread along with one-worker) - probably I am wrong as I did not tested all these cases, but want to make sure that we cover all cases.

Reply:
There's a description [here](https://gitlab.cern.ch/geant4/geant4/-/tree/master/source/run): `G4RUN_MANAGER_TYPE` is used when the run manager type is `Default`, and the `Only` extension will prevent the `G4FORCE_RUN_MANAGER_TYPE` environment variable from overriding the specified type.